### DISCUSSION TITLE
Multi-tenancy out-of-the-box

### DISCUSSION THREAD

**Discussion by Chris2Brooks** [directus/directus#2687] [Archive]
Hey all,

Thanks for the awesome framework you folks are building! Looking forward to Directus 9! The UI looks amazing!

I just want to check whether Directus 9 will support multi-tenancy out-of-the-box?
I'm playing with Directus 8 to see how I can offer a SAAS backend for my clients (tenants) where they can create multiple staff roles with different user permissions. Staff should be able to see other staff content (but not edit if not admin). 

If not, is there a way to make this work in Directus 8/9?

I would love to hear how anyone else is approaching similar needs.

**Comment by rijkvanzanten** [comment]
Heya! 

Directus has multi-tenancy in that you can create different projects that each have their own database. Within each project you have the ability to create different roles to manage specific access.

There is no "single-database-multi-tenancy" where you can have something like nested roles to separate "tenants" and have sub-roles for groups within these tenants if that's what you're referring to.

We have a plan in the works to update the permissions system to be rule based (see https://github.com/directus/api/issues/511) which should make the latter possible as well.

**Reply by Chris2Brooks** [reply]
Mmmm, that might work for me

> Directus has multi-tenancy in that you can create different projects that each have their own database. Within each project you have the ability to create different roles to manage specific access.

☝️ This might actually be sufficient for my needs at this stage.

Just to be sure: I can create a project per client and each project can have me, the client, and the clients' employees. With admin permission for the client and read/ limited access for employees?

Is this correct?

**Reply by rijkvanzanten** [reply]
That's correct! Sounds like you'd have three roles: admin (exists by default), staff, and employees (names up to you of course). You can then set the staff/employees roles to have less permissions for read/write in the system

**Comment by visualiz3** [comment]
hi @rijkvanzanten , 
firstly i would like to say this framework is really awesome, keep it up man.

any ideas on how would you go about building "single-database-multi-tenancy" using current version of directus? 

For example if we have vendors & consumers but we want to extend with different data & is it still possible to use directus authentication.

Thanks for your time!

**Comment by tvld** [comment]
@rijkvanzanten  So you would suggest several projects for one single user to achieve multi-tenancy? Would that mean that each user need to create a new login account per project? I suppose we can automate it with a step in between through a "server middleware", but it seems a bit cumbersome to update the same user fields and passwords all the time?

_Edit: Oops, it seems that you need to duplicate even the whole database... So for us: 2500 user groups means 2500 databases? Seriously ? I am sure I am missing something here... newbie ;-) )))))_

**Reply by rijkvanzanten** [reply]
It totally depends on what you're trying to achieve. You can setup a different project per tenant, assuming every tenant is totally scoped and doesn't interact with each other, or you can setup multiple user roles where each role is a single tenant. From what you're saying, it sounds like

> So for us: 2500 user groups means 2500 ~databases~ user roles

**Comment by tvld** [comment]
We have, i suppose, one of the most common SAAS models:
- _users_ are member of one or more _user-groups_
- users can have a different role in each group (owner in one group, editor in another,..)
- _users_ can only see/act, depending on their role, on _items_ owned by the user-group

We expect >2000 user groups and >20.000 users.  So, if I understand your vision, in Directus the approach is to solve this with the user roles?

**Reply by benhaynes** [reply]
Are those Roles/Users for the CMS, or for your project?

Also, these permissions are dynamic, so you can set conditions based on the current user or their role. In that way, you can avoid having to create additional roles... but this all depends on the specifics of what you're trying to achieve.

In the end, our move to "Dynamic Permissions" is not mutually exclusive from Users being able to belong to multiple roles. But for now (due to our real-world constraints of being OSS), we're focusing on granular permissions first.

**Reply by tvld** [reply]
>Are those Roles/Users for the CMS, or for your project?

The users are our project users. (they are marketing agents with their clients)

> these permissions are dynamic, so you can set conditions based on the current user or their role

That I don't understand... You are saying you can handle dynamic permissions already in v8? And we can set these per item ?

**Reply by benhaynes** [reply]
In v8 you can _sort of_ do this, but setting permissions based on the status field (which forces that to be a system field of sorts), with options for `mine` and `role`. In v9, we're simply removing the proprietary side of this, and allowing admins to set permissions based on and condition/filter... for example:

**Articles: READ** — IF `owner` = `SYSTEM_CURRENT_USER` OR `date_published` > `SYSTEM_NOW`

**Reply by benhaynes** [reply]
https://docs.directus.io/guides/permissions.html

**Reply by tvld** [reply]
Got that. But it is all new: Directus came under our attention only last Monday and we are running the api now on three dev stations, one test and one pre-production (great!). So far, the user group stuff seems the last major stumble block, or hopefully only a steep learning curve. After this weekend we hope to decide.

You are saying we can start with v8 and then tackle our "SAAS user group" approach, when v9 is released with "condition/filters" ?

**Reply by ghost** [reply]
I've got a very similar use case in mind. @tvld did you find an approach that works well? 

I'm tinkering with the idea hijacking Roles and making them into user-groups.

I would add M2O relationships on my relevant tables that tie back to the role id.  On my article table:  article.role => directus_roles.id.  This would let assign ownership of the article to that user-group.

And then each Role would have custom item permissions for each relevant collection. Here is a Read example:

{
    "role": {
        "_eq": "$CURRENT_ROLE"
    }
}

This would restrict access to items in that collection to users with that Role.

I just need to figure out the syntax for automatically setting article.role. I couldn't find an example of the syntax used for a Field Preset.   Just use: "$CURRENT_ROLE"  ?

@benhaynes is that along the lines of what you were envisioning for doing multi-tenancy within a single project on Directus 9?

**Reply by tvld** [reply]
@jrc2289 Sorry, but we ended up making our own backend with FeathersJs and are happy ... So I cant help you here)

**Reply by j-petty** [reply]
I'm a big fan of Directus but the new "Permission" based single-database multi-tenancy approach has me a at a bit of a loss. How is that easier that the traditional approach of separate Projects per tenant?

Sure you could setup User Roles for each tenant and assign the tenant's users to the role so they only see their resources. But that seems like a lot more maintenance overhead for an administrator, given you need to remember to update the role's permissions every time you create a new Collection, for example.

What's worse though is the experience as a super-user, someone who is responsible for managing Directus across tenants. This basically becomes impossible using the proposed Permissions approach. Obviously super-users require access to see and edit everything. But lots of the time they will want to view resources specific to a single tenant.

I'd like to understand better how other people are implementing single-database multi-tenancy in Directus 9. To me it seems the underlying data structure is still compatible with separate Projects per tenant, which would be my prefered outcome. However the UI doesn't support that currently so far as I can tell because there is no way of switching between Projects.

**Reply by benhaynes** [reply]
Using permissions is one option for multi-tenancy, but there's nothing stopping you from using projects for each tenant. The only change we've made is that instead of one installation managing multiple tenants, each tenant would have its own project instance. This is crucial for keeping custom logic, extensions, and other configuration isolated. It also removes significant amount of noise needed for managing multiple projects from the code base.

It's very much possible to manage a fleet of projects using kubernetes or other methods. The only thing potentially missing is a way to switch between projects from the App UI... but that's not always desired in multi-tenant situations anyway.

Also, we have a strict 80/20 rule, and I'm not sure that 80% of our user base requires multi-tenancy out of the box. Therefore, it's better to have this be a custom wrapper solution.

**Comment by tvld** [comment]


**Reply by benhaynes** [reply]
Haha, the new GH discussions are odd... but it'd be ideal to keep the conversation threaded within the original posts. 😉

**Reply by benhaynes** [reply]
Yeah, that makes sense... v9 will be in beta in about a month, and that would be the earliest you can reliably test to make sure it fits your needs. But it's an improvement on v8 with no back-tracking, so you should be fine.

**Reply by tvld** [reply]
Ok, we are not sure fully where we getting into. But we promise after we succeed, we will write a blog post on how to use Directus v9 for SAAS ;-)

**Reply by vvovkata** [reply]
> Ok, we are not sure fully where we getting into. But we promise after we succeed, we will write a blog post on how to use Directus v9 for SAAS ;-)

Hello dear,
I was searching for a blog post or any tutorial about how to use Directus 9 for SAAS but couldn't find one.  In the version 9 release post there is very little info regarding the multi-tenancy options.

**Reply by sahilpohare** [reply]
> Ok, we are not sure fully where we getting into. But we promise after we succeed, we will write a blog post on how to use Directus v9 for SAAS ;-)

hey @NightlyFlux 
Any updates on the blog? I am trying to build a multi-tenant crypto SAAS with Directus. Also is it possible for me to create my own auth adapter?

**Reply by oachoor** [reply]
Hello, is there any update on this?
I'm still trying to figure out the **Role Scoping** approach as described [here](https://docs.directus.io/concepts/projects/#multitenancy), especially how to scope the permissions within a single project 🤔 

I've set the following permission-rule to the collections:
```
{
    "owner": {
        "_eq": "$CURRENT_USER"
    }
}
```

I think that's not enough, since it's only restricted to one user but not an organisation (or a user-group).
Maybe I'm missing another rule such as this?
```
{
    "owner": {
        "_belongs_to": "$USER_GROUP"
    }
}
```

Also files are shared among all users, I couldn't find how to set the permission-rule on **File Library** level, is that possible?

**Reply by benhaynes** [reply]
Did you try `$CURRENT_ROLE`?

https://docs.directus.io/reference/filter-rules/#dynamic-variables

**Reply by oachoor** [reply]
@benhaynes works 👌 

``` 
{
    "my_many_to_one_role_field": {
        "_eq": "$CURRENT_ROLE"
    }
}
```

**Reply by staminna** [reply]
@oachoor @benhaynes is it possible to have multiple people, some collaborating, some not with different permissions per role? I would like to know if it is possible to have at least a per-role admin, or a way to launch a docker instance in parallel for each group of people, enhancing security. Something like user equals to $CURRENT_USER OR "User belongs to same Role ( as Creator)"  
<img width="598" alt="Captura de ecrã 2022-10-11, às 15 20 03" src="https://user-images.githubusercontent.com/515180/195193783-3922bd1b-6088-4ae7-a809-01df6edbece7.png">

**Reply by rijkvanzanten** [reply]
> @oachoor @benhaynes is it possible to have multiple people, some collaborating, some not with different permissions per role? I would like to know if it is possible to have at least a per-role admin, or a way to launch a docker instance in parallel for each group of people, enhancing security, dynamically.

Yes, that's possible, though it depends on your exact business requirements on how to configure that exactly of course. If you run multiple copies of Directus, you'd probably run those copies against different databases, in which case everything would be isolated for everybody 👍🏻

**Reply by staminna** [reply]
@rijkvanzanten Awesome! Is there advice for Docker or a link you can share for setting this up like this? Because running two copies makes double the effort of every maintenance and deployment task.

**Reply by rijkvanzanten** [reply]
That's getting into advanced devops land quick 👀 We put together https://directus.cloud so people don't have to deal with that, so that'd be my first recommendation 🙂
